cudamapper/src/matcher.cu

```diff
@@ -8,107 +8,236 @@
 * license agreement from NVIDIA CORPORATION is strictly prohibited.
 */
 
+#include <algorithm>
+#include <memory>
 #include "matcher.hpp"
+#include <cudautils/cudautils.hpp>
 
 namespace claragenomics {
 
-/*    typedef struct Anchor{
-        read_id_t query_read_id_;
-        read_id_t target_read_id_;
-        position_in_read_t query_position_in_read_;
-        position_in_read_t target_position_in_read_;
-    } Anchor;*/
-
-/*    __global__ void matcher(const position_in_read_t* const positions_in_read_d,
-                            const read_id_t* const read_ids_d,
-                            const ArrayBlock* const blocks_belonging_to_reads_d,
-                            const ArrayBlock* const sketch_elements_with_same_representation_to_check_d,
-                            const ArrayBlock* const sketch_elements_with_same_representation_and_read_d,
-                            Anchor** const anchors_for_read_d,
-                            std:uint32_t* const total_anchors_for_read_d) {
-        extern __shared__ position_in_read_t query_positions_in_read[]; // number of elements == blockDim.x
+    // TODO: move to cudautils?
+    /// \brief creates a unique pointer to device memory which gets deallocated automatically
+    ///
+    /// \param num_of_elems number of elements to allocate
+    ///
+    /// \return unique pointer to allocated memory
+    template<typename T>
+    std::unique_ptr<T, void(*)(T*)> make_unique_cuda_malloc(std::size_t num_of_elems) {
+        T* tmp_ptr_d = nullptr;
+        CGA_CU_CHECK_ERR(cudaMalloc((void**)&tmp_ptr_d, num_of_elems*sizeof(T)));
+        std::unique_ptr<T, void(*)(T*)> u_ptr_d(tmp_ptr_d, [](T* p) {CGA_CU_CHECK_ERR(cudaFree(p));}); // tmp_prt_d's ownership transfered to u_ptr_d
+        return std::move(u_ptr_d);
+    }
+
+    /// \brief Generates anchors for all reads
+    ///
+    /// Each thread block works on one query read. read_id_to_pointer_arrays_section_d points sections of read_id_to_sketch_elements_d and read_id_to_sketch_elements_to_check_d.
+    /// Kernel matches sketch elements that have the same representation. These sketch elements are pointed to by read_id_to_sketch_elements_d and read_id_to_sketch_elements_to_check_d.
+    ///
+    /// Kernel has an extern shared memory array which should be at least as large as the largest number of sketch elements with the same representation in any block
+    ///
+    /// Anchors are grouped by query read id and within that by representation (both in increasing order).
+    /// Assume q0p4t2p8 means anchor of read id 0 at position 4 and read id 2 at position 8.
+    /// Assume read 0 has 30 sketch elements with certain representation, read 1 40 and read 2 50.
+    /// Anchors for read 0 as query and that represtnation looks like this:
+    /// q0p0t1p0, q0p0t1p1 .. q0p0t1p39, q0p0t2p0, q0p0t2p1 ... q0p0t2p49, q0p1t1p0, q0p1t1p1 ... q0p1t1p39, q0p1t2p0 .. q0p1t2p49, q0p2p1p0 ...
```

Comment:
don't understand the 3 ts in the last element `q0p2p1p0`

Reply:
fixed, see above